File: samples_harder/sample.cs
class SampleHarder {

    int value;

    SampleHarder(int v) {
        if (v < 0) {
            throw new ArgumentException("negative");
        }
        value = v;
    }

    int NestedTryCatch(int a, int b) {
        int result;
        try {
            try {
                if (b == 0) {
                    throw new DivideByZeroException("zero");
                }
                result = a / b;
            } catch (DivideByZeroException) {
                if (a > 0) {
                    result = a;
                } else {
                    throw new InvalidOperationException("both bad");
                }
            }
            result += 1;
        } catch (InvalidOperationException) {
            result = -999;
        } finally {
            Console.WriteLine("outer finally");
        }
        return result;
    }

    int TryInLoop(int[] items) {
        int sum = 0;
        foreach (int item in items) {
            try {
                if (item < 0) {
                    throw new ArgumentException("neg");
                }
                sum += 100 / item;
            } catch (DivideByZeroException) {
                sum += -1;
                if (sum < -10) {
                    break;
                }
            } catch (ArgumentException) {
                continue;
            } finally {
                Console.WriteLine("processed");
            }
        }
        return sum;
    }

    int SwitchWithLoops(int[] data) {
        int state = 0;
        int output = 0;

        for (int i = 0; i < data.Length; i++) {
            switch (state) {
                case 0:
                    if (data[i] > 0) {
                        state = 1;
                    } else if (data[i] < 0) {
                        state = 2;
                    } else {
                        continue;
                    }
                    output++;
                    break;
                case 1:
                    for (int j = 0; j < data[i]; j++) {
                        output += j;
                        if (output > 100) {
                            return output;
                        }
                    }
                    state = 0;
                    break;
                case 2:
                    if (data[i] == -99) {
                        return -1;
                    }
                    output -= data[i];
                    state = 0;
                    break;
            }
        }
        return output;
    }

    void MultiGoto(int a, int b) {
        int result = 0;

        if (a < 0) {
            goto error_a;
        }
        if (b < 0) {
            goto error_b;
        }
        if (a == b) {
            goto equal;
        }

        result = a * b;
        goto done;

    error_a:
        result = -1;
        goto done;

    error_b:
        result = -2;
        goto done;

    equal:
        result = a * a;

    done:
        Console.WriteLine(result);
    }

    int DeepNesting(int[] items) {
        int total = 0;
        int i = 0;
        while (i < items.Length) {
            if (items[i] > 0) {
                for (int j = 0; j < items[i]; j++) {
                    if (j % 2 == 0) {
                        try {
                            if (items[i] / (j + 1) > 10) {
                                throw new OverflowException("too big");
                            }
                            total += items[i] / (j + 1);
                        } catch (OverflowException) {
                            continue;
                        }
                    } else {
                        total += 1;
                    }
                }
            } else if (items[i] == 0) {
                i++;
                continue;
            } else {
                return total;
            }
            i++;
        }
        return total;
    }

    int SwitchFallthrough(int code) {
        int result = 0;
        switch (code) {
            case 1:
            case 2:
            case 3:
                result = code * 10;
                break;
            case 4:
                result = 40;
                goto case 5;
            case 5:
                result += 50;
                break;
            case 6:
                if (code > 5) {
                    result = 60;
                    break;
                }
                goto case 7;
            case 7:
                result = 70;
                break;
            default:
                result = -1;
                break;
        }
        return result;
    }

    void NestedDoWhile(int limit) {
        int i = 0;
        do {
            int j = 0;
            do {
                if (i == j) {
                    j++;
                    continue;
                }
                if (i + j > limit * 2) {
                    break;
                }
                j++;
            } while (j < limit);
            i++;
            if (i == 3) {
                continue;
            }
        } while (i < limit);
    }

    void UsingWithTry(string path) {
        try {
            using (var reader = new StreamReader(path)) {
                string line;
                while ((line = reader.ReadLine()) != null) {
                    if (line.StartsWith("#")) {
                        continue;
                    }
                    if (line == "END") {
                        break;
                    }
                    Console.WriteLine(line);
                }
            }
        } catch (IOException) {
            Console.WriteLine("IO error");
        } finally {
            Console.WriteLine("done");
        }
    }

    void Run() {
        NestedTryCatch(10, 0);
        NestedTryCatch(-5, 0);
        int[] items = {1, 0, -1, 2, 3};
        TryInLoop(items);
        SwitchWithLoops(items);
        MultiGoto(3, -1);
        DeepNesting(items);
        SwitchFallthrough(4);
        NestedDoWhile(4);
        UsingWithTry("test.txt");
    }
}

File: samples_harder/sample.cpp
class Processor {
public:
    int state;

    Processor(int initial) {
        if (initial < 0) {
            throw -1;
        }
        state = initial;
    }

    int process(int items[], int n) {
        int result = 0;
        for (int i = 0; i < n; i++) {
            try {
                if (items[i] < 0) {
                    throw items[i];
                }
                result += items[i];
            } catch (int e) {
                if (e == -1) {
                    continue;
                }
                result = 0;
                break;
            }
        }
        return result;
    }

    int nested_try(int a, int b) {
        int result;
        try {
            try {
                if (b == 0) {
                    throw -1;
                }
                result = a / b;
            } catch (int inner) {
                if (a > 0) {
                    result = a;
                } else {
                    throw -2;
                }
            }
            result += 1;
        } catch (int outer) {
            result = -999;
        }
        return result;
    }
};

int classify_and_sum(int data[], int n) {
    int pos_sum = 0;
    int neg_count = 0;

    for (int i = 0; i < n; i++) {
        switch (data[i] > 0 ? 1 : (data[i] < 0 ? -1 : 0)) {
            case 1:
                pos_sum += data[i];
                if (pos_sum > 1000) {
                    return pos_sum;
                }
                break;
            case -1:
                neg_count++;
                if (neg_count > 5) {
                    goto bail;
                }
                break;
            case 0:
                continue;
        }
    }

    if (neg_count == 0) {
        return pos_sum;
    }
    return pos_sum - neg_count;

bail:
    return -1;
}

int deep_nesting(int matrix[][10], int rows, int cols) {
    int total = 0;
    for (int i = 0; i < rows; i++) {
        if (i % 2 == 0) {
            for (int j = 0; j < cols; j++) {
                if (matrix[i][j] > 0) {
                    while (matrix[i][j] > 10) {
                        matrix[i][j] /= 2;
                        if (matrix[i][j] == 7) {
                            break;
                        }
                    }
                    total += matrix[i][j];
                } else if (matrix[i][j] == 0) {
                    continue;
                } else {
                    total--;
                }
            }
        } else {
            for (int j = cols - 1; j >= 0; j--) {
                total += matrix[i][j];
            }
        }
    }
    return total;
}

void multi_loop_break(int items[], int n) {
    int i = 0;
    while (i < n) {
        for (int j = 0; j < n; j++) {
            if (items[j] == items[i]) {
                if (i != j) {
                    goto found_dup;
                }
                continue;
            }
        }
        i++;
    }
    return;

found_dup:
    items[i] = -1;
    return;
}

int state_machine_v2(int inputs[], int n) {
    int state = 0;
    int output = 0;

    for (int i = 0; i < n; i++) {
        switch (state) {
            case 0:
                if (inputs[i] == 1) {
                    state = 1;
                } else if (inputs[i] == 2) {
                    state = 2;
                } else {
                    continue;
                }
                output++;
                break;
            case 1:
                switch (inputs[i]) {
                    case 0:
                        state = 0;
                        break;
                    case 3:
                        state = 2;
                        output += 10;
                        break;
                    default:
                        output++;
                        break;
                }
                break;
            case 2:
                if (inputs[i] < 0) {
                    return output;
                }
                state = 0;
                break;
        }
    }
    return output;
}

int main() {
    Processor p(1);
    int items[] = {1, -1, 2, -3, 4};
    p.process(items, 5);
    p.nested_try(10, 0);

    classify_and_sum(items, 5);

    int mat[3][10] = {{1,2,3},{4,5,6},{7,8,9}};
    deep_nesting(mat, 3, 3);

    multi_loop_break(items, 5);

    int inputs[] = {1, 3, 0, 2, -1};
    state_machine_v2(inputs, 5);
    return 0;
}
